beacon_node/eth2-libp2p/src/rpc/protocol.rs

```
@@ -1,269 +1,318 @@
+        .to_vec()
+    }
 }
 
-impl UpgradeInfo for RPCEvent {
-    type Info = &'static [u8];
-    type InfoIter = iter::Once<Self::Info>;
+/* Inbound upgrade */
 
-    #[inline]
-    fn protocol_info(&self) -> Self::InfoIter {
-        iter::once(b"/eth/serenity/rpc/1.0.0")
-    }
-}
+// The inbound protocol reads the request, decodes it and returns the stream to the protocol
+// handler to respond to once ready.
 
-type FnDecodeRPCEvent = fn(Vec<u8>, ()) -> Result<RPCEvent, DecodeError>;
+pub type InboundOutput<TSocket> = (RPCRequest, InboundFramed<TSocket>);
+pub type InboundFramed<TSocket> = Framed<upgrade::Negotiated<TSocket>, InboundCodec>;
+type FnAndThen<TSocket> = fn(
+    (Option<RPCRequest>, InboundFramed<TSocket>),
+) -> FutureResult<InboundOutput<TSocket>, RPCError>;
+type FnMapErr<TSocket> = fn(timeout::Error<(RPCError, InboundFramed<TSocket>)>) -> RPCError;
 
 impl<TSocket> InboundUpgrade<TSocket> for RPCProtocol
 where
     TSocket: AsyncRead + AsyncWrite,
 {
-    type Output = RPCEvent;
-    type Error = DecodeError;
-    type Future = upgrade::ReadOneThen<upgrade::Negotiated<TSocket>, (), FnDecodeRPCEvent>;
-
-    fn upgrade_inbound(self, socket: upgrade::Negotiated<TSocket>, _: Self::Info) -> Self::Future {
-        upgrade::read_one_then(socket, MAX_READ_SIZE, (), |packet, ()| Ok(decode(packet)?))
+    type Output = InboundOutput<TSocket>;
+    type Error = RPCError;
+
+    type Future = future::AndThen<
+        future::MapErr<
+            timeout::Timeout<stream::StreamFuture<InboundFramed<TSocket>>>,
+            FnMapErr<TSocket>,
+        >,
+        FutureResult<InboundOutput<TSocket>, RPCError>,
+        FnAndThen<TSocket>,
+    >;
+
+    fn upgrade_inbound(
+        self,
+        socket: upgrade::Negotiated<TSocket>,
+        protocol: RawProtocolId,
+    ) -> Self::Future {
+        // TODO: Verify this
+        let protocol_id =
+            ProtocolId::from_bytes(&protocol).expect("Can decode all supported protocols");
```

Seems like a place to return an error, is that useful here? Not sure if there's some error-ful `Future` possible?

---

If this fails, then there is a coding error. I want this to fail loudly if we can't decode the protocols that we support.